### PR TITLE
Port release_90 patches to LLVM 10 monorepo

### PR DESCRIPTION
**Pull request by bryanpkc**
This pull request cherry-picks all relevant patches from the `release_90` branches in https://github.com/flang-compiler/flang-driver and https://github.com/flang-compiler/llvm, adapts them to fit the monorepo layout, and adjusts the Classic Flang code to let it coexist with the new Flang. To build the code, run CMake with `-DLLVM_ENABLE_CLASSIC_FLANG=on`.

See https://github.com/Huawei-PTLab/classic-flang-llvm/wiki for an example build script.

### CODE REVIEW

**Comment by kiranchandramohan** [comment]
@bryanpkc Is the CI (results above) running from your repository? Is it expected to fail?

**Comment by bryanpkc** [comment]
@kiranchandramohan I think those CI jobs came from the GitHub Actions that are configured for your repository:

https://github.com/flang-compiler/classic-flang-llvm-project/actions

These CI jobs are probably incorrect or irrelevant. For our purpose, a useful CI job should build llvm-project with `-DLLVM_ENABLE_CLASSIC_FLANG=ON`.

**Comment by bryanpkc** [comment]
I found a bug in my branch; I will push a fix shortly.

**Comment by RichBarton-Arm** [comment]
Classic flang had no pre-commit CI on any repositories before so these were not pre-existing CI configs. Presumably you made a fork of llvm/llvm-project to make flang-compiler/classic-flang-llvm-project? Is it possible that these were somehow inherited from llvm/llvm-project when the fork was made?

**Comment by bryanpkc** [comment]
> Classic flang had no pre-commit CI on any repositories before so these were not pre-existing CI configs. Presumably you made a fork of llvm/llvm-project to make flang-compiler/classic-flang-llvm-project? Is it possible that these were somehow inherited from llvm/llvm-project when the fork was made?

The master branch in classic-flang-llvm-project does not contain a `.github/workflows/` directory, however, the release_100 branch does. The directory is inherited from the release/10.x branch of llvm/llvm-project; I did not add the directory or configure any GitHub Action through the web UI.

It's not a bad idea to have some CI jobs to ensure that this repo is functionally correct, and especially can still build Clang normally.

**Comment by bryanpkc** [comment]
I have fixed this pull request to allow normal Clang builds and regression tests to succeed, except for the two tests mentioned in issue #2. Ideally we should fix those tests too so that we have a clean branch.

This port adds a `USE_CLASSIC_FLANG` macro to Clang (a28a4fa8259adba8048769391fec3b5218cf47f1), in order to resolve some conflicts between the classic Flang modifications and the F18-related changes in release/10.x. Please let me know if that is acceptable. Most other commits are straightforward cherry-pick of the corresponding patches from flang-compiler/flang-driver and flang-compiler/llvm.

**Comment by kiranchandramohan** [comment]
Was able to build flang with this monorepo using the build instructions provided. Copying the build instructions below (has trivial modifications to work with this monorepo). All tests in flang passes on Aarch64, two tests in llvm debug fails as @bryanpkc mentions.

I will go through the code next.

```
#!/bin/bash

buildtype=Release
installdir=$PWD/install

if [ ! -d classic-flang-llvm-project ]; then
  git clone https://github.com/flang-compiler/classic-flang-llvm-project.git
  pushd classic-flang-llvm-project
  git fetch origin pull/1/head:mono10
  git checkout mono10
  popd
fi
if [ ! -d flang ]; then
  git clone https://github.com/flang-compiler/flang
fi

mkdir -p build/llvm
pushd build/llvm
cmake -G Ninja \
  -DCMAKE_BUILD_TYPE=$buildtype \
  -DCMAKE_INSTALL_PREFIX=$installdir \
  -DCMAKE_C_COMPILER=gcc \
  -DCMAKE_CXX_COMPILER=g++ \
  -DLLVM_ENABLE_PROJECTS="clang;openmp" \
  -DLLVM_ENABLE_CLASSIC_FLANG=on \
  -DLLVM_ENABLE_ASSERTIONS=on \
  -DLLVM_OPTIMIZED_TABLEGEN=on \
  -DLLVM_TARGETS_TO_BUILD="X86;AArch64" \
  ../../classic-flang-llvm-project/llvm || exit 1
nice -n 10 ninja install || exit 1
popd

if ! llvm_lit="$(realpath --canonicalize-existing $PWD/build/llvm/bin/llvm-lit 2>/dev/null)" || \
    [ ! -x "$llvm_lit" ]; then
  echo "no usable llvm-lit"
  exit 1
fi

mkdir -p build/libpgmath
pushd build/libpgmath
cmake -G Ninja \
  -DCMAKE_BUILD_TYPE=$buildtype \
  -DCMAKE_INSTALL_PREFIX=$installdir \
  -DCMAKE_C_COMPILER=gcc \
  -DCMAKE_CXX_COMPILER=g++ \
  -DLLVM_CONFIG=$installdir/bin/llvm-config \
  -DLLVM_TARGETS_TO_BUILD="X86;AArch64" \
  -DLIBPGMATH_LLVM_LIT_EXECUTABLE="$llvm_lit" \
  ../../flang/runtime/libpgmath || exit 1
nice -n 10 ninja check-libpgmath install || exit 1
popd

mkdir -p build/flang
pushd build/flang
cmake -Wno-dev -G 'Unix Makefiles' \
  -DCMAKE_BUILD_TYPE=$buildtype \
  -DCMAKE_INSTALL_PREFIX=$installdir \
  -DCMAKE_C_COMPILER=$installdir/bin/clang \
  -DCMAKE_CXX_COMPILER=$installdir/bin/clang++ \
  -DCMAKE_Fortran_COMPILER=$installdir/bin/flang \
  -DLLVM_CONFIG=$installdir/bin/llvm-config \
  -DLLVM_TARGETS_TO_BUILD="X86;AArch64" \
  -DLLVM_EXTERNAL_LIT="$llvm_lit" \
  -DFLANG_LLVM_EXTENSIONS=ON \
  ../../flang || exit 1
nice -n 10 make -j48 install || exit 1
nice -n 10 make check-all || exit 1
popd

echo "SUCCESS"
```

File: clang/test/Driver/autocomplete.c
Comment (kiranchandramohan) [inline clang/test/Driver/autocomplete.c:75]: Would adding PGMATH to check be better here?
Or can this be not currently tested due to conditional compilation with FLANG_LLVM_EXTENSIONS?

File: clang/lib/Frontend/InitPreprocessor.cpp
Comment (kiranchandramohan) [inline clang/lib/Frontend/InitPreprocessor.cpp:162]: Why are these functions removed?

**Review by kiranchandramohan — COMMENTED** [review]
Two quick questions.

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
Comment (peterwaller-arm) [inline llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp]: Nit: trailing space on this line
```suggestion
    }
```

File: clang/test/Driver/autocomplete.c
Comment (bryanpkc) [reply clang/test/Driver/autocomplete.c:75]: Yes, I changed this test precisely because of the conditional compilation with `FLANG_LLVM_EXTENSIONS`.

If we merge `FLANG_LLVM_EXTENSIONS` with `USE_CLASSIC_FLANG`, then we could create a variant of this test, and distinguish between two with the `classic_flang` llvm-lit feature tag. However, it is also not very nice to have to maintain two variants of this test.

If llvm-lit supported conditional `RUN` lines, it would solve this problem. I could look into modifying the `RUN` lines to make `flang` indicate whether it is classic or not, *and then* choose the appropriate `FileCheck` command. This approach is quite hacky and requires additional changes to the driver, so it is not very appealing to me.

BTW I wasn't sure whether there are dependencies on `FLANG_LLVM_EXTENSIONS` that I'm not aware of. That's why I did not go ahead and merge the two macros into one.

File: clang/lib/Frontend/InitPreprocessor.cpp
Comment (bryanpkc) [reply clang/lib/Frontend/InitPreprocessor.cpp:162]: They were moved to `include/clang/Frontend/Utils.h` in flang-compiler/flang-driver@c51f89679135f84675f492d560ec5535c2000cfe, presumably because the functions are needed in `lib/Driver/ToolChains/Flang.cpp`.

**Comment by kiranchandramohan** [comment]
> I have fixed this pull request to allow normal Clang builds and regression tests to succeed, except for the two tests mentioned in issue #2. Ideally we should fix those tests too so that we have a clean branch.

I notice that the build instructions you have do not use the FLANG_LLVM_EXTENSIONS flag while building flang. If you use this flag then there will be DIGlobalVariables without the name field. This happens because there is some code (see link below) in lldebug.cpp which removes names of artificial variables.
https://github.com/flang-compiler/flang/blob/d99e82b764a6622e31bec31e61a80072f5c64291/tools/flang2/flang2exe/lldebug.cpp#L529 

The simplest testcase to reproduce the issue is,
```
module IEEE_ARITHMETIC
  use, intrinsic :: iso_c_binding
end module
```

**Comment by bryanpkc** [comment]
> I notice that the build instructions you have do not use the FLANG_LLVM_EXTENSIONS flag while building flang. If you use this flag then there will be DIGlobalVariables without the name field. This happens because there is some code (see link below) in lldebug.cpp which removes names of artificial variables.

I based my build instructions on the [official build instructions](https://github.com/flang-compiler/flang/wiki/Building-Flang) which also did not mention `FLANG_LLVM_EXTENSIONS`. Let me try to look at the `DIGlobalVariable` problem a bit further.

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (peterwaller-arm) [inline clang/lib/Driver/ToolChains/Clang.cpp]:
```suggestion
      if (val.equals("all")) {
```

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (peterwaller-arm) [inline clang/lib/Driver/ToolChains/Clang.cpp]:
```suggestion
      } else if (val.equals("inline") || val.equals("vect")) {
```

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (peterwaller-arm) [inline clang/lib/Driver/ToolChains/Clang.cpp]:
```suggestion
      if (val.equals("all")) {
```

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (peterwaller-arm) [inline clang/lib/Driver/ToolChains/Clang.cpp]:
```suggestion
      } else if (val.equals("inline") || val.equals("vect")) {
```

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
Comment (peterwaller-arm) [inline llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp]: There is a misaligned brace on the bottom here (aligned with `for`, should be aligned with `if`). The top `for` and the `if` should both have braces as they have multiline bodies.

```suggestion
      if (const DIFortranArrayType *AT = dyn_cast<DIFortranArrayType>(
              static_cast<const Metadata *>(DIV->getType()))) {
        for (const DINode *S : AT->getElements()) {
          if (const DIFortranSubrange *FS = dyn_cast<DIFortranSubrange>(S)) {
            if (const DIVariable *LBV = FS->getLowerBound())
              VariableInDependentType[LBV] = AT;
            if (const DIVariable *UBV = FS->getUpperBound())
              VariableInDependentType[UBV] = AT;
          }
        }
      }
```

File: clang/lib/Driver/ToolChain.cpp
Comment (peterwaller-arm) [inline clang/lib/Driver/ToolChain.cpp]:
```suggestion
  if (useOpenMP) {
```

File: clang/lib/Driver/ToolChain.cpp
Comment (peterwaller-arm) [inline clang/lib/Driver/ToolChain.cpp]:
```suggestion
  if (staticFlangLibs) {
```

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (peterwaller-arm) [inline clang/lib/Driver/ToolChains/Clang.cpp]:
```suggestion
  if (Args.getLastArg(options::OPT_fveclib))
```

File: clang/CMakeLists.txt
Comment (peterwaller-arm) [inline clang/CMakeLists.txt]: I think it makes sense to roll this into LLVM_ENABLE_CLASSIC_FLANG.

**Comment by bryanpkc** [comment]
@kiranchandramohan I believe it is best to combine `FLANG_LLVM_EXTENSIONS` and `USE_CLASSIC_FLANG` into one macro, e.g. `CLASSIC_FLANG_LLVM_EXTENSIONS`, and use this to switch between the two definitions of `DIGlobalVariable`, along with other Flang-specific features such as the frontend changes and `PGMATH`. We could then exclude the two failing unit tests when testing Classic Flang. I can update the branch if you agree that this is the right approach.

File: clang/include/clang/Basic/MacroBuilder.h
Comment (peterwaller-arm) [inline clang/include/clang/Basic/MacroBuilder.h:26]: This suggestion fixes warnings I see:

```suggestion
  MacroBuilder(raw_ostream &Output) : Out(Output) {}
  virtual ~MacroBuilder() {}
```

Building with clang-8, I get a screen full of warnings of the form:

```
class MacroBuilder {
      ^
1 warning generated.
[576/625] Building CXX object tools/clang/lib/Basic/CMakeFiles/obj.clangBasic.dir/Targets.cpp.o
In file included from /home/petwal01/classic-flang-llvm-project/clang/lib/Basic/Targets.cpp:14:
In file included from /home/petwal01/classic-flang-llvm-project/clang/lib/Basic/Targets.h:19:
/home/petwal01/classic-flang-llvm-project/clang/include/clang/Basic/MacroBuilder.h:23:7: warning: 'clang::MacroBuilder' has virtual functions but non-virtual destructor [-Wnon-virtual-dtor]
class MacroBuilder {
      ^
1 warning generated.
```

My cmake line looks like so:
```
CC=clang-8 CXX=clang++-8 cmake ../llvm -GNinja -DLLVM_USE_LINKER=lld -DLLVM_ENABLE_PROJECTS='clang' -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_ASSERTIONS=On -DLLVM_ENABLE_CLASSIC_FLANG=On
```

File: clang/lib/Driver/ToolChain.cpp
Comment (peterwaller-arm) [inline clang/lib/Driver/ToolChain.cpp:289]: I think this could be unified into getFlang(), having that one return tools::FlangFrontend if USE_CLASSIC_FLANG.

File: clang/lib/Driver/ToolChains/Flang1.cpp
Comment (peterwaller-arm) [inline clang/lib/Driver/ToolChains/Flang1.cpp]: How about calling this "ClassicFlang" rather than "Flang1"?

Flang1.cpp => ClassicFlang.cpp

File: clang/lib/Driver/ToolChains/Flang1.h
Comment (peterwaller-arm) [inline clang/lib/Driver/ToolChains/Flang1.h]: Perhaps this could be called `clang::driver::tools::ClassicFlang` to distinguish it from `clang::driver::tools::Flang` (and the filename ClassicFlang.h, .cpp, etc as well).

**Review by peterwaller-arm — CHANGES_REQUESTED** [review]
As someone who has been involved in monorepo and driver work elsewhere, this looks broadly good to me. I've reviewed this by studying the diff-of-the-diff vs the release_90 branches on `flang-compiler/flang-driver` and `flang-compiler/llvm`.

I think it's possible to simplify the driver, removing the FlangFrontend and instead using the Compiler phase. I've [published my preliminary patch](https://github.com/peterwaller-arm/classic-flang-llvm-project/commit/9078d6c4c8b7c65ea4adfb4102e276a5d54c8c1d), which I will submit for review after this PR has landed, or you're welcome to incorporate those changes here if you feel that is appropriate. I don't mind rebasing as needed for any changes you want to make or include in this patch.

I've highlighted some small issues. I don't think fixing them all needs to hold this PR back.

* I'd like to see the warning fixed, please.
* I think merging the two `#define`s into one is a good idea and have a personal preference for `USE_CLASSIC_FLANG`, but I'm sure other names would be fine too.

File: clang/lib/Driver/ToolChain.cpp
Comment (peterwaller-arm) [reply clang/lib/Driver/ToolChain.cpp:289]: (For reference, that's what I did in my patch [here](https://github.com/peterwaller-arm/classic-flang-llvm-project/commit/9078d6c4c8b7c65ea4adfb4102e276a5d54c8c1d#diff-099203948b69b8f6547254ea9599f840R266)). You don't need to do that in your patch unless you want to.

File: clang/test/Driver/autocomplete.c
Comment (bryanpkc) [reply clang/test/Driver/autocomplete.c:75]: I have followed @peterwaller-arm's advice and merged the two macros into one `ENABLE_CLASSIC_FLANG`.

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
Comment (bryanpkc) [reply llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp]: Fixed.

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (bryanpkc) [reply clang/lib/Driver/ToolChains/Clang.cpp]: All spacing issues in this file have been fixed.

File: clang/CMakeLists.txt
Comment (bryanpkc) [reply clang/CMakeLists.txt]: Done. Setting `LLVM_ENABLE_CLASSIC_FLANG=ON` will now define the `ENABLE_CLASSIC_FLANG` macro, which replaces both `USE_CLASSIC_FLANG` and `FLANG_LLVM_EXTENSIONS`.

File: clang/include/clang/Basic/MacroBuilder.h
Comment (bryanpkc) [reply clang/include/clang/Basic/MacroBuilder.h:26]: Applied your fix.

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
Comment (bryanpkc) [reply llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp]: Fixed.

File: clang/lib/Driver/ToolChain.cpp
Comment (bryanpkc) [reply clang/lib/Driver/ToolChain.cpp]: The spacing issues in this file have been fixed. I also got rid of the braces around single-line if-blocks.

File: clang/lib/Driver/ToolChain.cpp
Comment (bryanpkc) [reply clang/lib/Driver/ToolChain.cpp:289]: I am going to leave this change to your patch. It makes a lot of sense. Thanks.

File: clang/lib/Driver/ToolChains/Flang1.cpp
Comment (bryanpkc) [reply clang/lib/Driver/ToolChains/Flang1.cpp]: I have renamed the classes and files according to your suggestions.

File: clang/lib/Driver/ToolChains/Flang1.h
Comment (bryanpkc) [reply clang/lib/Driver/ToolChains/Flang1.h]: Done.

**Review by peterwaller-arm — APPROVED** [review]
Looks good to me, thanks!

**Review by kiranchandramohan — APPROVED** [review]
LGTM.
Builds and passes make check-all (llvm and flang) with a Release build on a AArch64 machine.

**Comment by kiranchandramohan** [comment]
This now needs approval from @gklimowicz and @shivaramaarao.

**Comment by gklimowicz** [comment]
I'm working on this today. It may not be done until Wednesday.

**Review by shivaramaarao — APPROVED** [review]
LGTM. build passed and also the tests passed.

**Review by gklimowicz — APPROVED** [review]
These changes for classic-flang-llvm-project look fine to me. I have run across 2 failures in `check-clang` on OpenPOWER, though. The first is in `f90_correct/fc30`. The second is in `f90_correct/ls07`.

These failures do not occur on LLVM 10 x86_64 or aarch64.

After digging into these failures, they appear to be related to how LLVM 10 on Power is handling negative zeros during the computation of COMPLEX multiplies. The two tests fail when they generate `-0.0` for the imaginary part of a COMPLEX number, when the test results are expecting `0.0`.

These need to be investigated further. There's a slight chance these are related to the driver changes in the pull request. It seems more likely to be on the LLVM 10 code-gen side.

I see no reason at this time to hold up approval of this pull request, though.

**Comment by bryanpkc** [comment]
I have pushed flang-compiler/flang#927 to exclude the test failures in f90_correct and created an issue flang-compiler/flang#928 to track their fix.

**Comment by bryanpkc** [comment]
@kiranchandramohan I don't have merge privilege in this repository. Could you merge this PR, or give me the access please? I think the PR is good to go now that we have a workaround (even if it hasn't been merged) for the test failures.

**Comment by gklimowicz** [comment]
@bryanpkc I am adding you to the organization, and then will make sure that you have merge privileges.

**Comment by kiranchandramohan** [comment]
@bryanpkc I hope @gklimowicz has given you commit access and u can merge now. If not then please let me know.

**Comment by bryanpkc** [comment]
@kiranchandramohan @gklimowicz has added me to the organization but I don't have merge privilege for this repository yet.

**Comment by kiranchandramohan** [comment]
OK @bryanpkc, i can merge. Is merge (no rebase or squash) the best action here?

**Comment by bryanpkc** [comment]
Please rebase if necessary, but do not squash. Keeping the historical commits separate make them easier to rebase to future release branches. Thank you.

**Comment by kiranchandramohan** [comment]
Flang works on llvm-10 now. Thanks @bryanpkc for your help.

**Comment by bryanpkc** [comment]
Thanks very much @kiranchandramohan.

**Comment by bryanpkc** [comment]
@peterwaller-arm Could you submit a pull request for your patch (peterwaller-arm/classic-flang-llvm-project@9078d6c4c8b7)?

**Comment by banach-space** [comment]
@bryanpkc As discussed in the Flang call yesterday (7/10/20), feel free to cherry-pick that patch and submit it on behalf of Peter. We discussed this with @peterwaller-arm offline and decided that that would be the preferred way.

Peter is currently on a different project and I'm actively working on the driver for the LLVM Flang. If you need reviewers, I volunteer :)

**Comment by peterwaller-arm** [comment]
I've published a rebase on https://github.com/peterwaller-arm/classic-flang-llvm-project/tree/simplify-driver, that's https://github.com/peterwaller-arm/classic-flang-llvm-project/commit/10dacac072ded415449e7350c7ba67c2b500f576. It would be great if someone can pick it up and run with it.

It has an issue: currently the classic_flang.f95 test fails, because the driver thinks that flang is capable of emitting assembly. I think what's going on is that ToolSelector::getTool is combining flang+llvm here: https://github.com/peterwaller-arm/classic-flang-llvm-project/blob/10dacac072ded415449e7350c7ba67c2b500f576/clang/lib/Driver/Driver.cpp#L4110-L4113

As far as I know that is the only issue, but it needs testing with flang and I don't have time to dive in to this for a bit.

**Pull request merged**
